### ISSUE TITLE
how to express  10 dimensional categorical variables

### ISSUE THREAD

**Issue by zdx3578** [openai/InfoGAN#10]
how to  express  10 dimensional categorical variables

code:
latent_spec = [
(Uniform(62), False),
(Categorical(10), True),
(Uniform(1, fix_std=True), True),
(Uniform(1, fix_std=True), True),
]
is for mnist ,
this is not enough,
in paper:
MNIST, we choose to model the latent codes with one categorical code, c1 ⇠ Cat(K = 10, p = 0.1), which can model discontinuous variation in data, and two continuous codes that can capture variations that are continuous in nature: c2 , c3 ⇠ Unif ( 1, 1).

but what to express: Street View House Number (SVHN
we make use of four 10 dimensional categorical variables and two uniform continuous variables as latent codes.

CelebA
In this dataset, we model the latent variation as 10 uniform categorical variables, each of dimension 10.

**Comment by zdx3578** [comment]
if isinstance(dist, Gaussian):
                assert dist.dim == 1, "Only dim=1 is currently supported"
                c_vals = []
                for idx in xrange(10):
                    c_vals.extend([-1.0 + idx * 2.0 / 9] * 10)
                c_vals.extend([0.] * (self.batch_size - 100))
                vary_cat = np.asarray(c_vals, dtype=np.float32).reshape((-1, 1))
                cur_cat = np.copy(fixed_cat)
                cur_cat[:, offset:offset+1] = vary_cat
                offset += 1
            elif isinstance(dist, Categorical):
                lookup = np.eye(dist.dim, dtype=np.float32)
                cat_ids = []
                for idx in xrange(10):
                    cat_ids.extend([idx] * 10)
                cat_ids.extend([0] * (self.batch_size - 100))
                cur_cat = np.copy(fixed_cat)
                cur_cat[:, offset:offset+dist.dim] = lookup[cat_ids]
                offset += dist.dim
            elif isinstance(dist, Bernoulli):
                assert dist.dim == 1, "Only dim=1 is currently supported"

**Comment by zdx3578** [comment]
embedding_dim = 100

    latent_spec = [
        (Uniform(64), False),
        (Categorical(32), True),
    ]
    con_latent_spec = [
        (LatentGaussian(embedding_dim), True)
    ]


https://github.com/RutgersHan/InfoGAN/blob/dev_auto/launchers/generate_images.py

good!

**Comment by zdx3578** [comment]
C.3 CelebA
The network architectures are shown in Table 3. The discriminator D and the recognition network Q shares most of the network. For this task, we use 10 ten-dimensional categorical code and 128 noise variables, resulting in a concatenated dimension of 228.

is

    latent_spec = [
        (Uniform(128), False),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
        (Categorical(10), True),
    ]


but how this??

C.5 Chairs
The network architectures are shown in Table 6. The discriminator D and the recognition network Q shares the same network, and only have separate output units at the last layer. For this task, we use 1 continuous latent code, **3 discrete latent codes (each with dimension 20)**, and 128 noise variables, so the input to the generator has dimension 189.

**Comment by NHDaly** [comment]
The above latent_spec worked okay for me.

```python
c3_celebA_latent_spec = [
    (Uniform(128), False),  # Noise
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
]
c3_celebA_image_size = 32
```

Can you elaborate a bit more in words what you're having problems with? I'm not sure I understand what's not working for you.

**Comment by zdx3578** [comment]
i think celebA config is ok,  now question is about  how to config C.5 Chairs as below

The network architectures are shown in Table 6. The discriminator D and the recognition network Q shares the same network, and only have separate output units at the last layer. For this task, we use 1 continuous latent code, **3 discrete latent codes (each with dimension 20)**, and 128 noise variables, so the input to the generator has dimension 189.

3 discrete latent codes (each with dimension 20) but visual code is below:
elif isinstance(dist, Bernoulli):
assert dist.dim == 1, "Only dim=1 is currently supported"


so how to config c.5 chairs,and how to change visual code ? or visual code wont change??

**Comment by NHDaly** [comment]
> how to config C.5 Chairs as below

I might be misunderstanding, but it seems like 

> For this task, we use 1 continuous latent code, **3 discrete latent codes (each with dimension 20)**, and 128 noise variables, so the input to the generator has dimension 189.

would translate to the following latent_spec. That is, the continuous code is represented by `Uniform` and the discrete code is represented by `Categorical`:

```python
c5_chairs_latent_spec = [
    (Uniform(128), False),  # Noise
    (Uniform(1, fix_std=True), True),
    (Categorical(20), True),
    (Categorical(20), True),
    (Categorical(20), True),
]
c3_celebA_image_size = 32
```

I copied the `(Uniform(1, fix_std=True), True)` line from the [two continuous variables defined in run_mnist_exp.py](https://github.com/openai/InfoGAN/blob/master/launchers/run_mnist_exp.py#L39), which I believe represent the  "2 continuous latent codes" referenced from the MNIST section of the paper.


I'm not sure where you got the LatentGaussian from... I don't know if it's necessary? I haven't tried running the Chairs model at all.

**Comment by NHDaly** [comment]
That said, I am also very curious about the answer to this question:

> how to config '**separate configurations for each learned variation**' ?
> ![](https://cloud.githubusercontent.com/assets/382863/19032832/8cad2366-898e-11e6-8800-536520ee7581.png)

Does this mean that you ran the experiment multiple times with the same number of codes, but each of the codes tends to perform best for each of the provided settings?

**Comment by neocxi** [comment]
> That is, the continuous code is represented by Uniform and the discrete code is represented by Categorical:

This is correct. Thanks @NHDaly !


> Does this mean that you ran the experiment multiple times with the same number of codes, but each of the codes tends to perform best for each of the provided settings?

Yes, to better compare with previous supervised results, we select codes from multiple runs that are most similar to categories that previous method (DC-IGN) produces.

**Comment by zdx3578** [comment]
for @NHDaly  ref this  https://github.com/RutgersHan/InfoGAN/blob/dev_auto/launchers/run_flower_exp.py#L49

for @neocxi  
    self.reg_cont_latent_dist = Product([x for x in self.reg_latent_dist.dists if isinstance(x, Gaussian)])
    self.reg_disc_latent_dist = Product([x for x in self.reg_latent_dist.dists if isinstance(x, (Categorical, Bernoulli))])
Bernoulli is also discrete. where to use Bernoulli?

can @neocxi  give a  example show which para is according to    ![image](https://cloud.githubusercontent.com/assets/382863/19070925/2e3ccfd4-8a60-11e6-8df1-63fb9fd3c396.png)

**Comment by zdx3578** [comment]
A.2 INFOGAN TRAINING
To train the InfoGAN network described in Tbl. 1 on the 2D shapes dataset (Fig. 6), we followed the training paradigm described in Chen et al. (2016) with the following modifications. For the mutual information regularised latent code, we used 5 continuous variables ci sampled uniformly from ( 1,1). We used 5 noise variables zi, as we found that using a reduced number of noise variables improved the quality of generated samples for this dataset. To help stabilise training, we used the instance noise trick described in Shi et al. (2016), adding Gaussian noise to the discriminator inputs (0.2 standard deviation on images scaled to [ 1, 1]). We followed Radford et al. (2015) for the architecture of the convolutional layers, and used batch normalisation in all layers except the last in the generator and the first in the discriminator.
 
from 
 beta-VAE: LEARNING BASIC VISUAL CONCEPTS WITH A CONSTRAINED VARIATIONAL FRAMEWORK

**Comment by simonzhang0158** [comment]
I am wondering if there is any update about CelebA dataset. I have the same problem as @zdx3578 when I am trying to train CelebA. @zdx3578 Did you find a way to solve this? The paper you mentioned above I believe is the set up for 2D shapes dataset.

Any help will be appreciate.
